lib/mode.coffee

```diff
@@ -0,0 +1,234 @@
+    primary_span = input.spans.find (span) -> span.is_primary
+    continue unless primary_span
+    range = [
+      [primary_span.line_start - 1, primary_span.column_start - 1],
+      [primary_span.line_end - 1, primary_span.column_end - 1]
+    ]
+    input.level = 'error' if input == 'fatal error'
+    element =
+      type: input.level
+      message: input.message
+      file: primary_span.file_name
+      range: range
+      children: input.children
+    for span in input.spans
+      unless span.is_primary
+        element.children.push
+          message: span.label
+          range: [
+            [span.line_start - 1, span.column_start - 1],
+            [span.line_end - 1, span.column_end - 1]
+          ]
+    elements.push element
+  buildMessages elements, disabledWarnings
+
+parseJsonOutput = (output, disabledWarnings) ->
+  results = output.split('\n').map (message) ->
+    message = message.trim()
+    if message.startsWith '{'
+      JSON.parse message
+  .filter (m) -> m?
+  parseJsonMessages results, disabledWarnings
+
+buildMessages = (elements, disabledWarnings) ->
+  messages = []
+  lastMessage = null
+  for element in elements
+    switch element.type
+      when 'info', 'trace', 'note'
+        # Add only if there is a last message
+        if lastMessage
+          lastMessage.trace or= []
+          lastMessage.trace.push
+            type: "Trace"
+            text: element.message
+            filePath: element.file
+            range: element.range
+      when 'warning'
+        # If the message is warning and user enabled disabling warnings
+        # Check if this warning is disabled
+        if disabledWarnings and disabledWarnings.length > 0
+          messageIsDisabledLint = false
+          for disabledWarning in disabledWarnings
+            # Find a disabled lint in warning message
+            if element.message.indexOf(disabledWarning) >= 0
+              messageIsDisabledLint = true
+              lastMessage = null
+              break
+          if not messageIsDisabledLint
+            lastMessage = constructMessage "Warning", element
+            messages.push lastMessage
+        else
+          lastMessage = constructMessage "Warning" , element
+          messages.push lastMessage
+      when 'error', 'fatal error'
+        lastMessage = constructMessage "Error", element
+        messages.push lastMessage
+  return messages
+
+constructMessage = (type, element) ->
+  message =
+    type: type
+    text: element.message
+    filePath: element.file
+    range: element.range
+  # children exists only in JSON messages
+  if element.children
+    message.trace = []
+    for children in element.children
+      message.trace.push
+        type: "Trace"
+        text: children.message
+        filePath: element.file
+        range: children.range or element.range
+  message
+
+buildRustcArguments = (linter, paths) ->
+  [editingFile, cargoManifestPath] = paths
+  Promise.resolve().then () =>
```

Comment:
🐛 It looks like you are trying to use this as a way to return a Promise, you should just go directly there:
`new Promise (resolve, reject) =>`

I'd consider this an optional change though I guess, as it makes the CoffeeScript of the function even worse than it is from the start 😛.